lib/src/main/java/net/fellbaum/jemoji/EmojiManager.java

```diff
@@ -474,22 +470,63 @@ public static String replaceEmojis(final String text, final String replacementSt
     public static String replaceEmojis(final String text, Function<Emoji, String> replacementFunction, final Collection<Emoji> emojisToReplace) {
         if (isStringNullOrEmpty(text)) return "";
 
-        final int[] textCodePointsArray = text.codePoints().toArray();
-        final long textCodePointsLength = textCodePointsArray.length;
+        final StringBuilder sb = new StringBuilder(text.length());
 
-        final StringBuilder sb = new StringBuilder();
+        forEachEmoji(text, new VoidEmojiProcessor() {
+            private int currentCodepoint;
+
+            @Override
+            public void onCodepoint(int codepoint) {
+                currentCodepoint = codepoint;
+
+                sb.appendCodePoint(codepoint);
+            }
+
+            @Override
+            public FunctionResult<Void> onEmoji(int textIndex, Emoji emoji) {
+                sb.delete(sb.length() - Character.charCount(currentCodepoint), sb.length());
+
+                if (emojisToReplace.contains(emoji)) {
+                    sb.append(replacementFunction.apply(emoji));
+                } else {
+                    sb.append(emoji.getEmoji());
+                }
+
+                return FunctionResult.keepLooking();
+            }
+        });
+
+        return sb.toString();
+    }
+
+    /**
+     * Replaces all emojis in the text with the given replacement function.
+     *
+     * @param text                The text to replace emojis from.
+     * @param replacementFunction The replacement function.
+     * @param emojisToReplace     The emojis to replace.
+     * @return The text with all emojis replaced.
+     */
+    public static String replaceEmojis(final String text, Function<Emoji, String> replacementFunction, final Emoji... emojisToReplace) {
+        return replaceEmojis(text, replacementFunction, Arrays.asList(emojisToReplace));
+    }
+
+    private static <T> T forEachEmoji(final String text,
+                                      final EmojiProcessor<T> processor) {
+        final int[] textCodePointsArray = stringToCodePoints(text);
+        final long textCodePointsLength = textCodePointsArray.length;
```

Comment:
This is a hot spot, there's a way to get the code point from the string, which would save us from getting the entire array (except in worst case of course), but i couldn't figure out a way to get the correct index

Might be interesting though: https://stackoverflow.com/a/1527891 and `String#offsetByCodePoints`